enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/protocol/ProtocolInstallerRepository.java

```diff
@@ -49,16 +58,44 @@ private void addTo( Map<Protocol,ProtocolInstaller<O>> tempServerMap, ProtocolIn
         }
     }
 
-    public ProtocolInstaller installerFor( Protocol protocol )
+    public ProtocolInstaller<O> installerFor( ProtocolStack protocolStack )
     {
-        ProtocolInstaller protocolInstaller = installers.get( protocol );
-        if ( protocolInstaller == null )
+        ApplicationProtocol applicationProtocol = protocolStack.applicationProtocol();
+        ProtocolInstaller.Factory<O,?> protocolInstaller = installers.get( applicationProtocol );
+
+        // TODO need at least one modifier to frame
```

Comment:
I wonder if we actually always need the basic framing for POJOs. Snappy does some chunking and framing, but I suspect it might not work as we need it to. We need it to emit entire ByteBufs representing complete messages for the subsequent decoders. Perhaps this is the reason some tests are failing? From looking at the source code of SnappyFrameEncoder I think for example that sending a message above 32K will yield several ByteBuf chunks on the decoding side. I might be totally wrong of course as well.

Reply:
Mixing Snappy with basic framing breaks everything. I haven't tested Snappy with large messages.

Reply:
Hmm... it really shouldn't. I wonder if that is an issue of ordering in the pipeline? I think snappy should be the outermost layer and the basic framing within it, since snappy compresses a stream and the basic framing collects until it has a complete frame.